README.md
# node-red-contrib-sony-audio-control
A collection of Node-RED nodes to control Sony audio devices via the Sony Audio Control API.

---
**NOTE**

This package has been deprecated in favor of the new node collection [node-red-contrib-sony-audio](https://github.com/jensrossbach/node-red-contrib-sony-audio). This means the package is now in maintainance mode and will only get updates for severe bugs and security issues. There will not be any new features developed for these nodes.

---

With these nodes you can connect to your Sony audio devices and send control commands (like powering on/off, change volume etc.), retrieve information (like the current volume, the current value of a sound setting etc.) or get notified of certain events (like a change of the power state, a change of the mute state etc.).

![Sony Audio Control](images/sony_audio_control.png)

If you would like to propose a new feature or any kind of improvement or if you want to report a bug, don't hesitate to create an issue in the [issue tracker](https://github.com/jensrossbach/node-red-contrib-sony-audio-control/issues) of the repository on GitHub.

- [Documentation](#documentation)
  - [Request Node](#request-node)
    - [Input](#input)
    - [Outputs](#outputs)
  - [Event Node](#event-node)
     - [Configuration](#configuration-1)
     - [Output](#output)
  - [Control Node](#control-node)
    - [Configuration](#configuration-2)
    - [Input](#input-1)
    - [Outputs](#outputs-1)
  - [Filter Node](#filter-node)
    - [Configuration](#configuration-3)
    - [Input](#input-2)
    - [Outputs](#outputs-2)
- [License](#license)
- [Sony Legal Information](#sony-legal-information)
    - [Trademark](#trademark)
    - [License Audio Control API](#license-audio-control-api)

## Documentation
The node collection consists of four flow nodes and one configuration node. The configuration node stores the network address of your Sony audio device. You have to specify a host name (or IP address) and the port (by default, port 10000 is used).

### Request Node
The request node of the Node-RED Sony Audio Control node collection is the interface to the request/response API of Sony audio devices. It provides the full power of the Sony Audio Control API but requires some knowledge of the API when used stand-alone. Therefore it is recommended to combine a request node with a control node and/or a filter node. If you still want to use the node stand-alone (because maybe the required functionality is not available through the control node), please refer to the Sony [Audio Control API](https://developer.sony.com/develop/audio-control-api/api-references/api-overview-2) reference for further information.

![Request Node](images/request_node.png)

#### Configuration
In the configuration page of the node, you can choose a name (if no name is provided, the default name will be "request") and select the configuration node dedicated to your Sony audio device.

#### Input
The input message of the request node must contain the three properties `msg.service`, `msg.method` and `msg.version`. The service property refers to the API service which provides the desired method. The method property contains the request method of the API and the version property refers to the API version of the method. Additionally the `msg.payload` must contain the parameters of the request if applicable. It has to be an object containing the parameters as properties. There is no need to encapsulate the payload into an array, this is automatically done by the request node. For methods that do not have parameters, the `msg.payload` must be set to `null`.

#### Outputs
The request node provides two outputs. The first output contains the result of the request in the `msg.payload`. A message is only sent on success. The encapsulating array as coming from the device is already removed, so the payload is either an object with the result properties or an array in case a list of properties is returned. For the case that the request has no response parameters (empty result), the payload will be `null`. The second output sends a message with the error details in case the request failed. The information is provided in the `msg.payload` which is an object with the properties `error` containing the error code and `description` containing a human readable description of the error.

The error codes are defined by the Sony Audio Control API, however some node related error might also occur:

|Error Code|Description                                |
|----------|-------------------------------------------|
|32768     |Invalid node input                         |
|32769     |Invalid command (to the control node)      |
|32770     |Error in communication with the API service|

Both outputs additionally contain the properties `msg.service`, `msg.method` and `msg.version` corresponding to the request.

### Event Node
The event node of the Node-RED Sony Audio Control node collection is the interface to the notification API of Sony audio devices. It allows to subscribe for certain notification events of the different API services and provides the details of the notification on its output. The information provided at the output is in the format sent by the device and requires some knowledge of the API when used stand-alone. Therefore it is recommended to combine an event node with a filter node. If you still want to use the node stand-alone (because maybe the required filtering is not available through the filter node), please refer to the Sony [Audio Control API](https://developer.sony.com/develop/audio-control-api/api-references/api-overview-2) reference for further information.

![Event Node](images/event_node.png)

#### Configuration
In the configuration page of the node, you can choose a name (if no name is provided, the default name will be "event: _service_") and select the configuration node dedicated to your Sony audio device. Besides of this, you have to choose the service which provides the notification events you want to subscribe for. Depending on the selected service, a set of checkboxes appears where you can select the notification events to subscribe. If you want to get notified for events from multiple services, you have to create a dedicated event node for each service of interest.

#### Output
On the output a message is sent each time a subscribed notification event is fired by the device. It contains the parameters of the notification in the `msg.payload`. The encapsulating array as coming from the device is already removed, so the payload is either an object with the event properties or an array in case a list of properties is sent. For the case that the notification has no parameters (empty parameter set), the payload will be `null`. The message additionally contains the properties `msg.service`, `msg.method` and `msg.version` corresponding to the notification.

### Control Node
The control node of the Node-RED Sony Audio Control node collection supports the creation of requests to control a Sony audio device. The node itself is not able to communicate with the device and therefore not useable stand-alone. Rather its output must be connected to a request node in order to send requests to the device.

![Control Node](images/control_node.png)

#### Configuration
In the configuration page of the node, you can choose the name of the node (if no name is provided, the default name will be "control: _command_"). Additionally the command can be chosen which will be executed unless it is overridden by the input message. Depending on the selected command, different settings may be configured. Also here applies that parameters from the input message might (partly) override the chosen settings.

The following commands can be configured. The table shows also how they map to the commands that can be specified via the input message.

|Configuration         |Command         |Description                                                   |
|----------------------|----------------|--------------------------------------------------------------|
|Get Power Status      |getPowerStatus  |Retrieve current power status of the device                   |
|Power ON              |powerOn         |Switch on power of the device                                 |
|Power OFF             |powerOff        |Switch off power of the device (device goes into standby mode)|
|Get Source            |getSource       |Retrieve the currently active audio source on the device      |
|Set Source            |setSource       |Set the currently active audio source                         |
|Get Volume Information|getVolumeInfo   |Retrieve information about current volume and mute state      |
|Set Volume            |setVolume       |Set current volume                                            |
|Mute                  |mute            |Mute the audio output                                         |
|Unmute                |unmute          |Unmute the audio output                                       |
|Toggle Mute           |toggleMute      |Toggle between muted and unmuted audio output                 |
|Get Sound Settings    |getSoundSettings|Retrieve current sound settings on the device                 |
|Set Sound Settings    |setSoundSettings|Set sound settings on the device                              |
|Get Playback Modes    |getPlaybackModes|Retrieve current playback modes on the device                 |
|Set Playback Modes    |setPlaybackModes|Set playback modes on the device                              |
|Stop                  |stop            |Stop current audio playback                                   |
|Toggle Pause          |togglePause     |Toggle between paused and playing state                       |
|Skip Previous         |skipPrev        |Skip to previous or beginning of currently playing content    |
|Skip Next             |skipNext        |Skip to next content                                          |
|Scan Backward         |scanBackward    |Scan forward in currently playing content                     |
|Scan Forward          |scanForward     |Scan backward in currently playing content                    |

Below table shows the mapping from configuration settings to payload parameters which can be provided via the input message. See next chapter for more information about overwriting settings via the input message.

|Configuration                     |Payload       |
|----------------------------------|--------------|
|Source                            |type, source  |
|Port                              |port          |
|Volume                            |volume        |
|Relative Volume                   |relativeVolume|
|List of Sound Settings<sup>1</sup>|soundSettings |
|Setting                           |target        |
|List of Playback Modes<sup>2</sup>|modeSettings  |
|Mode                              |target        |
|Zone                              |zone          |

<sup>1</sup> The list of sound settings is a list where you can add new rows each representing a sound setting. You can choose the setting from the dropdown box and then depending on the selected setting, either turn the setting on or off, or select the setting's value from a second dropdown box.

<sup>2</sup> The list of playback modes is a list where you can add new rows each representing a playback mode. You can choose the mode type from the dropdown box and then select the mode's value from a second dropdown box.

#### Input
The input of the control node is used to trigger an action on one side and can be utilized to override the command and its settings (or a part of them) from the configurations page on the other side.

The command of the request can be overridden by specifying it in the `msg.command` argument of the input message. Command settings (for instance the volume of a `setVolume` command) can be overridden by adding them to the `msg.payload` argument of the input message. You can override all command settings or only some of them. All settings which are not specified in the input message will be taken from the node configuration. The `msg.payload` parameter must be an object with one property for each command setting.

The following properties are defined:

|Property      |Applicable to Command|Type        |Description                                                                                |
|--------------|----------------     |------------|-------------------------------------------------------------------------------------------|
|type<sup>1</sup>|setSource          |String      |The type of the audio source to be activated                                               |
|source<sup>1</sup>|setSource        |String      |The audio source to be activated                                                           |
|port<sup>1</sup>|setSource          |Number [1-9]|The port for an HDMI input                                                                 |
|volume        |setVolume            |Number      |The volume to set, can either be an absolute volume (>= 0) or a relative volume step (!= 0)|
|relativeVolume|setVolume            |Boolean     |True if the provided volume is a relative volume step, false otherwise                     |
|soundSettings<sup>2</sup>|setSoundSettings|Array |The sound settings to be activated                                                         |
|modeSettings<sup>3</sup>|setPlaybackModes|Array  |The playback modes to be activated                                                         |
|target<sup>4</sup>|getSoundSettings, getPlaybackModes|String|The type of sound setting or playback mode to retrieve                          |
|zone<sup>5</sup>|setSource, getSource, setVolume, getVolumeInfo, mute, unmute, toggleMute, stop, togglePause, skipPrev, skipNext, scanBackward, scanForward|Number [0-9]|The ouput zone of the device|

<sup>1</sup> The audio source can be specified via the properties `type`, `source` and `port`. The latter is only needed for HDMI sources. The following combinations are possible:

|type    |source  |port |
|--------|--------|-----|
|extInput|tv      |     |
|extInput|sat-catv|     |
|extInput|hdmi    |[1-9]|
|extInput|video   |     |
|extInput|sacd-cd |     |
|extInput|bd-dvd  |     |
|extInput|line    |     |
|extInput|btAudio |     |
|extInput|game    |     |
|extInput|source  |     |
|storage |usb1    |     |
|dlna    |music   |     |
|radio   |fm      |     |

<sup>2</sup> The sound settings can be specified via property `soundSettings` which is an array of objects each consisting of the properties `target` and `value`. The following combinations are possible:

|target      |value     |
|------------|----------|
|soundField  |off       |
|soundField  |standard  |
|soundField  |clearAudio|
|soundField  |music     |
|soundField  |movie     |
|soundField  |sports    |
|soundField  |game      |
|clearAudio  |off       |
|clearAudio  |on        |
|nightMode   |off       |
|nightMode   |on        |
|footballMode|off       |
|footballMode|on        |
|voice       |type1     |
|voice       |type2     |
|voice       |type3     |

<sup>3</sup> The playback modes can be specified via property `modeSettings` which is an array of objects each consisting of the properties `target` and `value`. The following combinations are possible:

|target      |value       |
|------------|------------|
|playType    |normal      |
|playType    |repeatAll   |
|playType    |repeatFolder|
|playType    |repeatTrack |
|playType    |shuffleAll  |
|repeatType  |off         |
|repeatType  |all         |
|repeatType  |folder      |
|repeatType  |track       |
|repeatType  |chapter     |
|shuffleType |off         |
|shuffleType |folder      |


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.2.2] - 2019-07-27
### Changed
- Marked package as deprecated in flavor of its successor, package [node-red-contrib-sony-audio](https://github.com/jensrossbach/node-red-contrib-sony-audio).

### Fixed
- Removed command "standby" as it didn't work (device goes into standby mode when issuing powerOff command).

## [1.2.1] - 2019-06-11
### Fixed
- Fixed reference error when sending two requests quickly one after another.
- Fixed wrong error codes in documentation part of [README.md](README.md).

## [1.2.0] - 2019-05-05
### Added
- Added support for playback control commands (stop, pause, next, previous, ...).
- Added support for playback modes (repeat, shuffle, ...).
- Added option _All_ to sound setting selection in control node.
- Added option _Any_ to sound setting filter in filter node.

### Changed
- Improved documentation and Sony legal information in [README.md](README.md).

## [1.1.2] - 2019-04-16
### Added
- Added status indicators for control and filter node showing configuration issues.
- Added notice in [README.md](README.md) to use issue tracker for feature requests and bug reports.

### Changed
- Updated screenshots for documentation.

### Fixed
- Fixed status indicators not disappearing on status change.

## [1.1.1] - 2019-04-14
### Fixed
- Aligned `msg.payload` format of request node input message with documentation.

> **NOTE**
>
> Although this is only a bugfix release, the payload format of the request node
> input message has been changed. This affects you only if you don't use the
> control node but use a custom node providing input to the request node. In
> this case you have to remove the encapsulating array braces around the actual
> content of the payload and use null as payload for methods without parameters.

## [1.1.0] - 2019-04-13
### Added
- Added setting "only if true" for powered and standby filters.
- Request and event node now show status indicators.

### Changed
- Incorporated documentation into main [README.md](README.md) file.
- Updated node screenshots for documentation.

## [1.0.1] - 2019-04-08
### Added
- Added a changelog.

### Changed
- Name of configuration nodes is now optional.
- Default name of configuration nodes is now the host name only (without port).
- Port property in configuration nodes is now mandatory.

### Fixed
- Fixed broken link to online documentation in [README.md](README.md) when viewing from Node-RED library.
- Removed outdated parts of documentation.
- Added documentation of missing functionality.

## [1.0.0] - 2019-04-07
Initial version and first published release.

nodes/sony-audio-control.js
/*
 * Copyright (c) 2019 Jens-Uwe Rossbach
 *
 * This code is licensed under the MIT License.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

module.exports = function(RED)
{
    function SonyAudioControlNode(config)
    {
        RED.nodes.createNode(this, config);

        this.name = config.name;
        this.command = config.command;
        this.volume = config.volume;
        this.relative = config.relative;
        this.source = config.source;
        this.zone = config.zone;
        this.port = config.port;
        this.soundSettings = config.soundsettings;
        this.soundTarget = config.soundtarget;
        this.modeSettings = config.modesettings;
        this.modeTarget = config.modetarget;

        function setPowerStatus(status)
        {
            return {service: "system",
                    method: "setPowerStatus",
                    version: "1.1",
                    payload: {status: status}};
        }

        function setAudioVolume(volume, relative = false, zone = 0)
        {
            return {service: "audio",
                    method: "setAudioVolume",
                    version: "1.1",
                    payload: {output: (zone > 0) ? "extOutput:zone?zone=" + zone : "",
                              volume: (relative && (volume > 0)) ? "+" + volume : volume.toString()}};
        }

        function setAudioMute(mute, zone = 0)
        {
            return {service: "audio",
                    method: "setAudioMute",
                    version: "1.1",
                    payload: {output: (zone > 0) ? "extOutput:zone?zone=" + zone : "",
                              mute: mute}};
        }

        function setSoundSettings(params)
        {
            return {service: "audio",
                    method: "setSoundSettings",
                    version: "1.1",
                    payload: {settings: params}};
        }

        function setPlaybackModeSettings(params)
        {
            return {service: "avContent",
                    method: "setPlaybackModeSettings",
                    version: "1.0",
                    payload: {settings: params}};
        }

        function setPlayContent(source, port = 0, zone = 0)
        {
            var uri = source;
            if ((source == "extInput:hdmi") && (port > 0))
            {
                uri += "?port=" + port;
            }

            return {service: "avContent",
                    method: "setPlayContent",
                    version: "1.2",
                    payload: {output: (zone > 0) ? "extOutput:zone?zone=" + zone : "",
                              uri: uri}};
        }

        function stopPlayingContent(zone = 0)
        {
            return {service: "avContent",
                    method: "stopPlayingContent",
                    version: "1.1",
                    payload: {output: (zone > 0) ? "extOutput:zone?zone=" + zone : ""}};
        }

        function pausePlayingContent(zone = 0)
        {
            return {service: "avContent",
                    method: "pausePlayingContent",
                    version: "1.1",
                    payload: {output: (zone > 0) ? "extOutput:zone?zone=" + zone : ""}};
        }

        function setPlayPreviousContent(zone = 0)
        {
            return {service: "avContent",
                    method: "setPlayPreviousContent",
                    version: "1.0",
                    payload: {output: (zone > 0) ? "extOutput:zone?zone=" + zone : ""}};
        }

        function setPlayNextContent(zone = 0)
        {
            return {service: "avContent",
                    method: "setPlayNextContent",
                    version: "1.0",
                    payload: {output: (zone > 0) ? "extOutput:zone?zone=" + zone : ""}};
        }

        function scanPlayingContent(fwd, zone = 0)
        {
            return {service: "avContent",
                    method: "scanPlayingContent",
                    version: "1.0",
                    payload: {direction: fwd ? "fwd" : "bwd",
                              output: (zone > 0) ? "extOutput:zone?zone=" + zone : ""}};
        }

        function getPowerStatus()
        {
            return {service: "system",
                    method: "getPowerStatus",
                    version: "1.1",
                    payload: null};
        }

        function getPlayingContentInfo(zone = 0)
        {
            return {service: "avContent",
                    method: "getPlayingContentInfo",
                    version: "1.2",
                    payload: {output: (zone > 0) ? "extOutput:zone?zone=" + zone : ""}};
        }

        function getVolumeInfo(zone = 0)
        {
            return {service: "audio",
                    method: "getVolumeInformation",
                    version: "1.1",
                    payload: {output: (zone > 0) ? "extOutput:zone?zone=" + zone : ""}};
        }

        function getSoundSettings(target)
        {
            return {service: "audio",
                    method: "getSoundSettings",
                    version: "1.1",
                    payload: {target: target}};
        }

        function getPlaybackModeSettings(target)
        {
            return {service: "avContent",
                    method: "getPlaybackModeSettings",
                    version: "1.0",
                    payload: {target: target}};
        }

        if ((this.command == "setSoundSettings") &&
            (this.soundSettings.length == 0))
        {
            this.status({fill: "yellow", shape: "dot", text: "configuration errors"});
        }
        else if ((this.command == "setPlaybackModes") &&
                 (this.modeSettings.length == 0))
        {
            this.status({fill: "yellow", shape: "dot", text: "configuration errors"});
        }
        else
        {
            this.status({});
        }

        this.on("input", function(msg)
        {
            var cmd = (typeof msg.command == "string") ? msg.command : this.command;

            switch (cmd)
            {
                case "powerOn":
                {
                    this.send(setPowerStatus("active"));
                    break;
                }
                case "powerOff":
                {
                    this.send(setPowerStatus("off"));
                    break;
                }
                case "setVolume":
                {
                    let args = {volume: this.volume,
                                relativeVolume: this.relative,
                                zone: this.zone};

                    if (typeof msg.payload == "object")
                    {
                        if (typeof msg.payload.volume == "number")
                        {
                            args.volume = msg.payload.volume;
                        }

                        if (typeof msg.payload.relativeVolume == "boolean")
                        {
                            args.relativeVolume = msg.payload.relativeVolume;
                        }

                        if (typeof msg.payload.zone == "number")
                        {
                            args.zone = msg.payload.zone;
                        }

                        if ((args.relativeVolume && (args.volume == 0)) ||
                            (!args.relativeVolume && (args.volume < 0)))
                        {
                            let errMsg = {method: "_passthrough", payload: {error: 32768, description: "Invalid node input"}};
                            this.send(errMsg);

                            break;
                        }
                    }

                    this.send(setAudioVolume(args.volume, args.relativeVolume, args.zone));
                    break;
                }
                case "mute":
                {
                    let args = {zone: this.zone};

                    if ((typeof msg.payload == "object") &&
                        (typeof msg.payload.zone == "number"))
                    {
                        args.zone = msg.payload.zone;
                    }

                    this.send(setAudioMute("on", args.zone));
                    break;
                }
                case "unmute":
                {
                    let args = {zone: this.zone};

                    if ((typeof msg.payload == "object") &&
                        (typeof msg.payload.zone == "number"))
                    {
                        args.zone = msg.payload.zone;
                    }

                    this.send(setAudioMute("off", args.zone));
                    break;
                }
                case "toggleMute":
                {
                    let args = {zone: this.zone};

                    if ((typeof msg.payload == "object") &&
                        (typeof msg.payload.zone == "number"))
                    {
                        args.zone = msg.payload.zone;
                    }

                    this.send(setAudioMute("toggle", args.zone));
                    break;
                }
                case "setSoundSettings":
                {
                    let args = {soundSettings: this.soundSettings};

                    if ((typeof msg.payload == "object") &&
                        Array.isArray(msg.payload.soundSettings))
                    {
                        args.soundSettings = msg.payload.soundSettings;
                    }

                    this.send(setSoundSettings(args.soundSettings));
                    break;
                }
                case "setSource":
                {
                    let args = {source: this.source,
                                port: this.port,
                                zone: this.zone};

                    if (typeof msg.payload == "object")
                    {
                        if ((typeof msg.payload.type == "string") &&
                            (typeof msg.payload.source == "string"))
                        {
                            args.source = msg.payload.type + ":" + msg.payload.source;
                        }

                        if (typeof msg.payload.port == "number")
                        {
                            args.port = msg.payload.port
                        }

                        if (typeof msg.payload.zone == "number")
                        {
                            args.zone = msg.payload.zone;
                        }
                    }

                    this.send(setPlayContent(args.source, args.port, args.zone));
                    break;
                }
                case "setPlaybackModes":
                {
                    let args = {modeSettings: this.modeSettings};

                    if ((typeof msg.payload == "object") &&
                        Array.isArray(msg.payload.modeSettings))
                    {
                        args.modeSettings = msg.payload.modeSettings;
                    }

                    this.send(setPlaybackModeSettings(args.modeSettings));
                    break;
                }
                case "stop":
                {
                    let args = {zone: this.zone};

                    if ((typeof msg.payload == "object") &&
                        (typeof msg.payload.zone == "number"))
                    {
                        args.zone = msg.payload.zone;
                    }

                    this.send(stopPlayingContent(args.zone));
                    break;
                }
                case "togglePause":
                {
                    let args = {zone: this.zone};

                    if ((typeof msg.payload == "object") &&
                        (typeof msg.payload.zone == "number"))
                    {
                        args.zone = msg.payload.zone;
                    }

                    this.send(pausePlayingContent(args.zone));
                    break;
                }
                case "skipPrev":
                {
                    let args = {zone: this.zone};

                    if ((typeof msg.payload == "object") &&
                        (typeof msg.payload.zone == "number"))
                    {
                        args.zone = msg.payload.zone;
                    }

                    this.send(setPlayPreviousContent(args.zone));
                    break;
                }
                case "skipNext":
                {
                    let args = {zone: this.zone};

                    if ((typeof msg.payload == "object") &&
                        (typeof msg.payload.zone == "number"))
                    {
                        args.zone = msg.payload.zone;
                    }

                    this.send(setPlayNextContent(args.zone));
                    break;
                }
                case "scanBackward":
                {
                    let args = {zone: this.zone};

                    if ((typeof msg.payload == "object") &&
                        (typeof msg.payload.zone == "number"))
                    {
                        args.zone = msg.payload.zone;
                    }

                    this.send(scanPlayingContent(false, args.zone));
                    break;
                }
                case "scanForward":
                {
                    let args = {zone: this.zone};

                    if ((typeof msg.payload == "object") &&
                        (typeof msg.payload.zone == "number"))
                    {
                        args.zone = msg.payload.zone;
                    }

                    this.send(scanPlayingContent(true, args.zone));
                    break;
                }
                case "getPowerStatus":
                {
                    this.send(getPowerStatus());
                    break;
                }
                case "getSource":
                {
                    let args = {zone: this.zone};

                    if ((typeof msg.payload == "object") &&
                        (typeof msg.payload.zone == "number"))
                    {
                        args.zone = msg.payload.zone;
                    }

                    this.send(getPlayingContentInfo(args.zone));
                    break;
                }
                case "getVolumeInfo":
                {
                    let args = {zone: this.zone};

                    if ((typeof msg.payload == "object") &&
                        (typeof msg.payload.zone == "number"))
                    {
                        args.zone = msg.payload.zone;
                    }

                    this.send(getVolumeInfo(args.zone));
                    break;
                }
                case "getSoundSettings":
                {
                    let args = {target: (this.soundTarget == "all") ? "" : this.soundTarget};

                    if ((typeof msg.payload == "object") &&
                        (typeof msg.payload.target == "string"))
                    {
                        args.target = msg.payload.target;
                    }

                    this.send(getSoundSettings(args.target));
                    break;
                }
                case "getPlaybackModes":
                {
                    let args = {target: (this.modeTarget == "all") ? "" : this.modeTarget};

                    if ((typeof msg.payload == "object") &&
                        (typeof msg.payload.target == "string"))
                    {
                        args.target = msg.payload.target;
                    }

                    this.send(getPlaybackModeSettings(args.target));
                    break;
                }
                default:
                {
                    let errMsg = {method: "_passthrough", payload: {error: 32769, description: "Invalid command"}};
                    this.send(errMsg);
                }
            }
        });
    }

    RED.nodes.registerType("sony-audio-control", SonyAudioControlNode);
}